File: kBUDGETApp/Sources/Shared/Services/CSVExportService.swift
import Foundation

protocol DataExporting {
    func generateCSV(for transactions: [Transaction]) throws -> URL
}

final class CSVExportService: DataExporting {
    private let fileManager: FileManager

    init(fileManager: FileManager = .default) {
        self.fileManager = fileManager
    }

    func generateCSV(for transactions: [Transaction]) throws -> URL {
        var rows: [String] = ["Date,Category,Original Amount,Original Currency,Amount CAD,Merchant,Note"]
        let formatter = ISO8601DateFormatter()

        for transaction in transactions {
            let dateString = formatter.string(from: transaction.date)
            let merchant = transaction.merchant?.replacingOccurrences(of: ",", with: " ") ?? ""
            let note = transaction.note?.replacingOccurrences(of: ",", with: " ") ?? ""
            let row = "\(dateString),\(transaction.category.name),\(transaction.amountOriginal),\(transaction.currencyCode),\(transaction.amountCAD),\(merchant),\(note)"
            rows.append(row)
        }

        let csvString = rows.joined(separator: "\n")
        let data = Data(csvString.utf8)
        let tempURL = fileManager.temporaryDirectory.appending(path: "kbudget-export-\(UUID().uuidString).csv")
        try data.write(to: tempURL)
        return tempURL
    }
}

File: kBUDGETReactNative/src/services/csvExportService.ts
import * as FileSystem from 'expo-file-system';
import { Transaction } from '@/models';

export interface DataExporter {
  exportTransactions(transactions: Transaction[]): Promise<string>;
}

export class CSVExportService implements DataExporter {
  async exportTransactions(transactions: Transaction[]): Promise<string> {
    const header = 'Date,Category,Original Amount,Original Currency,Amount CAD,Merchant,Note';
    const rows = transactions.map((transaction) => {
      const date = transaction.date;
      const merchant = (transaction.merchant ?? '').replace(/,/g, ' ');
      const note = (transaction.note ?? '').replace(/,/g, ' ');
      return `${date},${transaction.categoryId},${transaction.amountOriginal},${transaction.currencyCode},${transaction.amountCAD},${merchant},${note}`;
    });

    const content = [header, ...rows].join('\n');
    const fileUri = `${FileSystem.cacheDirectory}kbudget-${Date.now()}.csv`;
    await FileSystem.writeAsStringAsync(fileUri, content, { encoding: FileSystem.EncodingType.UTF8 });
    return fileUri;
  }
}
